### DISCUSSION TITLE
Dual Arm control

### DISCUSSION THREAD

**Discussion by mcaniot** [NVlabs/curobo#209] [Software Q&A]
Hi,
I'm using curobo with a dual arm configuration but i wonder how to manage to give goal on a specific end effector. For now, I m using the plan_single function. The only solution I found for controlling the other end effector is to provide the current position as goal to the end effector and place inside the link_poses the goal I want for the other end effector. It kinda work but i see some differences between controlling the left or the right gripper. Like it is not the same position error on the same, mirror, goal. Do you know if there are another solution ?

**Comment by balakumar-s** [comment]
If you want to move only one arm at a time, it's better to lock joints of the other arm when planning and switch locks to plan for the second arm.

**Reply by mcaniot** [reply]
meaning I need to have multiple configs files and cant have one for all my motion ? I cant change the end effector goal on the fly ?

**Reply by mcaniot** [reply]
Is it in the pipeline to be able to send dual goal instead of one ?

**Reply by bheijden** [reply]
@balakumar-s I am using a dual-arm setup (LEFT, RIGHT). In `robot_cfg` I set `ee_link` to `LEFT_ee`.

What is the recommended way to command the RIGHT arm while keeping the LEFT fixed? Should I:

1. keep two `robot_cfg` files with `ee_link` set per arm, or
2. declare both end effectors in one config, then switch the active `ee_link` at runtime?

Some questions the pop up about this "hot switching":
* Can I freeze the LEFT joints at runtime while planning for RIGHT, then unfreeze and swap sides, without restarting?
* Also, can you set the joints to a particular config while you have them defines as fixed in the `robot_cfg`?
* Do caches or kinematic solvers need to be rebuilt when switching? I'd like to avoid a lot of recompilation at runtime.

Some pointers or a very minimal example that shows how this switching would look would be super helpful! So how we can switch the  `ee_link` and toggle a fixed-joint set (at a particular configuration) efficiently

**Comment by peter-mitrano-bg** [comment]
This is also something I'm interested in. Specifically, I'd like to be able to define pose goals for both arms, and plan for all the joints simultaneously. It doesn't seem like one could do this without modifying core curobo functionality at the moment.

**Comment by balakumar-s** [comment]
You can send targets for multiple end-effectors. See this for an example: https://curobo.org/get_started/2b_isaacsim_examples.html#multi-arm-reacher

**Reply by peter-mitrano-bg** [reply]
Ah yes I should have been able to find that. That said, this API is a bit confusing. It seems to take the two goals in with different arguments in different formats (Pose, dict of Pose). I would have expected there to be a different goal type/function, rather than an additional optional argument where one has to be sure to only specify the goals for links that are not the "default" ee_link. Even having a "default" does not make sense for some robots, which the left/right sides are equally valid and you should always need to specify at planning time which should be used. Just my thoughts. Thanks for the quick response!

**Reply by peter-mitrano-bg** [reply]
One follow question I have is -- how do I specify constraints in this case? I modified my constrained planning example to add `link_poses={'other_arm': otherPose}` and it only worked if I disabled my constraints (e.g. `plan_config.pose_cost_metric = None`). Thanks!

**Reply by balakumar-s** [reply]
Constraints have currently been integrated only for the main ee_link. While constraints would work for additional links, we haven't setup the API. 

Regarding API choice on default ee_link and link_poses, link_poses was added much later in the project. I do agree it can be confusing. Ideally, we want to have a separate `MultiLinkMotionGen` class which should handle pose reaching for multiple links instead of overriding standard `MotionGen`. Currently, Multi-arm/Multi-link is an experimental feature and we will definitely revisit API choice once sufficient users find it useful.

**Reply by yunfanjiang** [reply]
> Constraints have currently been integrated only for the main ee_link. While constraints would work for additional links, we haven't setup the API.

Would like to follow up on this since I want my second arm to just reach partial poses. Do you have any estimation when this feature will be added? Or could you please suggest how to modify by myself? Thanks.

**Reply by bheijden** [reply]
Seconded. This feature would be valuable.

**Comment by DJT777** [comment]
I would also like to express my interest in more functionality for multi arm planning

**Comment by Alice-ly** [comment]
Hi, I want to follow this feature's progress as well, and I have figured out controlling two arms in one xrdf, but have not got valid solution on controlling two arms respectively due to different groups.

**Comment by sjauhri** [comment]
Hi @balakumar-s! I wanted to follow up on this too. Is there a way we can perform motion generation for a bimanual robot with collision constraints for both the end effectors?
